src/com/intellij/plugins/bodhi/pmd/PMDInvoker.java

```diff
@@ -68,33 +71,53 @@ public void runPMD(AnActionEvent actionEvent, String rule, boolean isCustomRuleS
         //Show the tool window
         PMDUtil.getProjectComponent(actionEvent).setupToolWindow();
 
+        Project project = actionEvent.getData(DataKeys.PROJECT);
+        PMDProjectComponent projectComponent = project.getComponent(PMDProjectComponent.class);
+
         List<File> files = new LinkedList<File>();
-        if (actionEvent.getPlace().equals(ActionPlaces.PROJECT_VIEW_POPUP)) {
+        if (actionEvent.getPlace().equals(ActionPlaces.PROJECT_VIEW_POPUP)
+                || actionEvent.getPlace().equals(ActionPlaces.SCOPE_VIEW_POPUP)
+                || actionEvent.getPlace().equals(ActionPlaces.CHANGES_VIEW_POPUP)
+                || actionEvent.getPlace().equals(ActionPlaces.MAIN_MENU)
+                ) {
+
             //If selected by right-click on file/folder (s)
-            VirtualFile[] selectedFiles = actionEvent.getData(DataKeys.VIRTUAL_FILE_ARRAY);
-            if (selectedFiles.length == 0) {
-                //toolWindow.displayErrorMessage("Please select a file to process first");
-                return;
+            VirtualFile[] selectedFiles;
+            if(actionEvent.getPlace().equals(ActionPlaces.CHANGES_VIEW_POPUP))
+            {
+                selectedFiles = VcsContextFactory.SERVICE.getInstance().createContextOn(actionEvent).getSelectedFiles();
             }
-            for (int i = 0; i < selectedFiles.length; i++) {
-                //Add all java files recursively
-                PMDUtil.listFiles(new File(selectedFiles[i].getPresentableUrl()), files,
-                        PMDUtil.createFileExtensionFilter("java"));
+            else if(actionEvent.getPlace().equals(ActionPlaces.MAIN_MENU))
+            {
+                VirtualFile[] contentRoots = ProjectRootManager.getInstance(project).getContentRoots();
+                selectedFiles = VfsUtil.getCommonAncestors(contentRoots);
             }
-        } else if (actionEvent.getPlace().equals("ChangesViewPopup")) {
-            VirtualFile[] selectedFiles = VcsContextFactory.SERVICE.getInstance().createContextOn(actionEvent).getSelectedFiles();
-            if (selectedFiles.length == 0) {
+            else
+            {
+                selectedFiles = actionEvent.getData(DataKeys.VIRTUAL_FILE_ARRAY);
+            }
+
+            if (selectedFiles == null || selectedFiles.length == 0) {
                 //toolWindow.displayErrorMessage("Please select a file to process first");
                 return;
             }
-            for (int i = 0; i < selectedFiles.length; i++) {
+            List<VirtualFileFilter> filters = new ArrayList<>();
```

Comment:
This requires Java language level 7 - currently we are using level 5  (to support old versions of Idea etc). Can you change this to:  `List<VirtualFileFilter> filters = new ArrayList<VirtualFileFilter>();`